### ISSUE TITLE
[WIIU] Regression: Plugging in certain USB HID devices disconnects the Wii U GamePad

### ISSUE THREAD

**Issue by vaguerant** [libretro/RetroArch#13309]
## Description

Following on from #13198. Keyboards are working again with #13308, but the other HID issues remain. In order of severity:

- the Wii U GamePad is disconnected by RetroArch if an unrecognized device is connected. This can occur either at launch if the devices are already inserted, or whenever the devices actually are inserted.
- certain unrecognized USB devices are detected as Generic SNES USB controllers, even though they aren't those. If a device **is** falsely recognized as a SNES USB controller, it doesn't seem to cause further issues, so fixing just this may make the above issue worse.

### Expected behavior

At worst, plugging in USB devices should just not work (for unsupported devices) or work (for supported devices) without disconnecting the GamePad or being identified in the RetroArch interface as Generic SNES USB controllers.

### Actual behavior
I've tested with four devices:
- **PowerWave Analog Game Controller**, some janky mid-2000s DInput Xbox knockoff pad
  - pops a notification, "Generic SNES USB controller (0/0) not configured". PowerWave controller does not work. GamePad still works.
- **THEC64 joystick**, a reproduction DInput Commodore 64 joystick
  - pops a notification, "Device disconnected from port #&#x2060;1." The notification *does not say "Wii U GamePad"*, which I believe a normal disconnection notice would; there is no mention of what the "device" being disconnected was. I suspect it's trying to disconnect the THEC64 joystick but it was never assigned a port because it's not recognized, and the result is it disconnects the GamePad instead. Controller does not work, as expected since it finds no match. On the upside, that means this controller is **not** ID'd as a Generic SNES USB controller.
- **Lenovo wired optical mouse** from some old computer or other
  - seems to do literally nothing. No notification pops and I don't think RetroArch on Wii U has any mouse support, so the mouse not actually working is expected behavior. No issues here.
- **Logitech K600 TV**, a wireless keyboard/mouse with dongle--in theory this is a proprietary Logitech Unifying device, but Unifying devices fall back on behaving as standard USB HID devices if they don't get a special init sequence from the Unifying driver, which Wii U obviously doesn't have, so I think this should be roughly equivalent to a wired keyboard and mouse plugged into a USB hub
  - pops a notification, "Device disconnected from port #&#x2060;1." The keyboard works perfectly, but the GamePad obviously doesn't work any more. The mouse, as expected, does nothing.

### Logs

I will note, I tried to log using `wiiu/net_listen.sh` as requested by @gblues, but only got output from Salamander; there's zero lines of log coming from the actual core. My RetroArch log level is set to **0 (Debug)** for both Frontend and Core. I can share the log if you really want it, but it stops at `2021-12-02T21:20:03.964885 [INFO] [Config]: Loading config from: "sd:/retroarch/retroarch.cfg".` There's really nothing there. Failing that, here's some regular RetroArch logs.

This first log is of me starting RetroArch, inserting the **PowerWave** controller (`vid: 0x8f0e pid: 0x300`), then the **THEC64** joystick (`vid: 0x591c pid: 0x2400`). Things of interest:
- it doesn't say it failed to find an entry for the PowerWave controller--presumably because it "matches" the Generic SNES USB controller entry?
- when THEC64 joystick is connected, it explicitly says it failed to find a match. This is when the Wii U GamePad stops working.
- RetroArch reports the name of the device as `TH`. Its actual name is `THEC64 Joystick`.
```
[INFO] RetroArch 1.9.13 (Git 121ca3a482)
[INFO] === Build =======================================
[INFO] Capabilities: 
[INFO] Built: Dec  2 2021
[INFO] Version: 1.9.13
[INFO] Git: 121ca3a482
[INFO] =================================================
[INFO] [Input]: Found input driver: "wiiu".
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] Version of libretro API: 1
[INFO] Compiled against API: 1
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Video @ 960x720
[INFO] [hid]: initializing
[INFO] [hid]: starting polling thread.
[INFO] [Input]: Found HID driver: "wiiu".
[INFO] [hid]: polling thread is starting
[INFO] [Joypad]: Found joypad driver: "wiiu".
[INFO] [Video]: Setting refresh rate to: 59.940 Hz.
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Video]: Found display server: null
[INFO] [Display]: Found display driver: "gx2".
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [SRAM]: SRAM will not be saved.
[INFO] [Playlist]: Loading history file: [sd:/retroarch/content_history.lpl].
[INFO] [Playlist]: Loading history file: [sd:/retroarch/content_music_history.lpl].
[INFO] [Playlist]: Loading favorites file: [sd:/retroarch/content_favorites.lpl].
[INFO]                 handle: 1919081
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0x8f0e
[INFO]                    pid: 0x300
[INFO]        interface_index: 0
[INFO]              sub_class: 0
[INFO]               protocol: 0
[INFO]     max_packet_size_rx: 8
[INFO]     max_packet_size_tx: 0
[INFO] connecting pad to slot 1
[INFO]                 handle: 1971588
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0x591c
[INFO]                    pid: 0x2400
[INFO]        interface_index: 0
[INFO]              sub_class: 0
[INFO]               protocol: 0
[INFO]     max_packet_size_rx: 8
[INFO]     max_packet_size_tx: 0
[INFO] Failed to find entry for vid: 0x591c, pid: 0x2400, name:  TH
[INFO] [Config]: Saving salamander config to: "sd:/retroarch/retroarch-salamander.cfg".
[INFO] [Config]: Saved new config to "sd:/retroarch/retroarch.cfg".
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 00 seconds.
[INFO] [Core]: Unloading core..
[INFO] [Core]: Unloading core symbols..
[INFO] [Video]: Average monitor Hz: 59.941258 Hz. (0.978 % frame time deviation, based on 2048 last samples).
[INFO] [Video]: Average monitor Hz: 59.941258 Hz. (0.978 % frame time deviation, based on 2048 last samples).
```
This next log is for the **Lenovo mouse**. Nothing very interesting happens. Does it poll the mouse four times? Doesn't really matter, no issues here.
```
[INFO] RetroArch 1.9.13 (Git 121ca3a482)
[INFO] === Build =======================================
[INFO] Capabilities: 
[INFO] Built: Dec  2 2021
[INFO] Version: 1.9.13
[INFO] Git: 121ca3a482
[INFO] =================================================
[INFO] [Input]: Found input driver: "wiiu".
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] Version of libretro API: 1
[INFO] Compiled against API: 1
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Video @ 960x720
[INFO] [hid]: initializing
[INFO] [hid]: starting polling thread.
[INFO] [Input]: Found HID driver: "wiiu".
[INFO] [hid]: polling thread is starting
[INFO] [Joypad]: Found joypad driver: "wiiu".
[INFO] [Video]: Setting refresh rate to: 59.940 Hz.
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Video]: Found display server: null
[INFO] [Display]: Found display driver: "gx2".
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [SRAM]: SRAM will not be saved.
[INFO] [Playlist]: Loading history file: [sd:/retroarch/content_history.lpl].
[INFO] [Playlist]: Loading history file: [sd:/retroarch/content_music_history.lpl].
[INFO] [Playlist]: Loading favorites file: [sd:/retroarch/content_favorites.lpl].
[INFO]                 handle: 1971596
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0xef17
[INFO]                    pid: 0x1960
[INFO]        interface_index: 0
[INFO]              sub_class: 1
[INFO]               protocol: 2
[INFO]     max_packet_size_rx: 4
[INFO]     max_packet_size_tx: 0
[INFO]                 handle: 1971598
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0xef17
[INFO]                    pid: 0x1960
[INFO]        interface_index: 0
[INFO]              sub_class: 1
[INFO]               protocol: 2
[INFO]     max_packet_size_rx: 4
[INFO]     max_packet_size_tx: 0
[INFO]                 handle: 1971600
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0xef17
[INFO]                    pid: 0x1960
[INFO]        interface_index: 0
[INFO]              sub_class: 1
[INFO]               protocol: 2
[INFO]     max_packet_size_rx: 4
[INFO]     max_packet_size_tx: 0
[INFO]                 handle: 1971602
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0xef17
[INFO]                    pid: 0x1960
[INFO]        interface_index: 0
[INFO]              sub_class: 1
[INFO]               protocol: 2
[INFO]     max_packet_size_rx: 4
[INFO]     max_packet_size_tx: 0
[INFO] [Config]: Saving salamander config to: "sd:/retroarch/retroarch-salamander.cfg".
[INFO] [Config]: Saved new config to "sd:/retroarch/retroarch.cfg".
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 00 seconds.
[INFO] [Core]: Unloading core..
[INFO] [Core]: Unloading core symbols..
[INFO] [Video]: Average monitor Hz: 59.941258 Hz. (0.958 % frame time deviation, based on 2048 last samples).
[INFO] [Video]: Average monitor Hz: 59.941258 Hz. (0.958 % frame time deviation, based on 2048 last samples).
```
Final log is for the **Logitech K600 TV**. Things of interest:
- The dongle reports three devices according to the `interface_index` lines. I assume a keyboard and mouse are two of those devices, no idea what the third would be. The storage for the configurable firmware maybe?
- RetroArch does explicitly fail to find an entry for the USB receiver (or perhaps just one of the three devices?). Like the THEC64 joystick, this also results in the GamePad disconnecting. Seems related.
```
[INFO] RetroArch 1.9.13 (Git 121ca3a482)
[INFO] === Build =======================================
[INFO] Capabilities: 
[INFO] Built: Dec  2 2021
[INFO] Version: 1.9.13
[INFO] Git: 121ca3a482
[INFO] =================================================
[INFO] [Input]: Found input driver: "wiiu".
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] Version of libretro API: 1
[INFO] Compiled against API: 1
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Video]: Video @ 960x720
[INFO] [hid]: initializing
[INFO] [hid]: starting polling thread.
[INFO] [Input]: Found HID driver: "wiiu".
[INFO] [hid]: polling thread is starting
[INFO] [Joypad]: Found joypad driver: "wiiu".
[INFO] [Video]: Setting refresh rate to: 59.940 Hz.
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Video]: Found display server: null
[INFO] [Display]: Found display driver: "gx2".
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [Font]: Using font rendering backend: stb-unicode.
[INFO] [SRAM]: SRAM will not be saved.
[INFO] [Playlist]: Loading history file: [sd:/retroarch/content_history.lpl].
[INFO] [Playlist]: Loading history file: [sd:/retroarch/content_music_history.lpl].
[INFO] [Playlist]: Loading favorites file: [sd:/retroarch/content_favorites.lpl].
[INFO]                 handle: 1971610
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0x6d04
[INFO]                    pid: 0x2bc5
[INFO]        interface_index: 0
[INFO]              sub_class: 1
[INFO]               protocol: 1
[INFO]     max_packet_size_rx: 8
[INFO]     max_packet_size_tx: 0
[INFO]                 handle: 1919083
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0x6d04
[INFO]                    pid: 0x2bc5
[INFO]        interface_index: 1
[INFO]              sub_class: 1
[INFO]               protocol: 2
[INFO]     max_packet_size_rx: 8
[INFO]     max_packet_size_tx: 0
[INFO]                 handle: 2063277
[INFO]   physical_device_inst: 0
[INFO]                    vid: 0x6d04
[INFO]                    pid: 0x2bc5
[INFO]        interface_index: 2
[INFO]              sub_class: 0
[INFO]               protocol: 0
[INFO]     max_packet_size_rx: 32
[INFO]     max_packet_size_tx: 0
[INFO] Failed to find entry for vid: 0x6d04, pid: 0x2bc5, name: USB Receiver
[INFO] [Config]: Saving salamander config to: "sd:/retroarch/retroarch-salamander.cfg".
[INFO] [Config]: Saved new config to "sd:/retroarch/retroarch.cfg".
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 00 seconds.
[INFO] [Core]: Unloading core..
[INFO] [Core]: Unloading core symbols..
[INFO] [Video]: Does not have enough samples for monitor refresh rate estimation. Requires to run for at least 4096 frames.
[INFO] [Video]: Does not have enough samples for monitor refresh rate estimation. Requires to run for at least 4096 frames.
```

### Steps to reproduce the bug

1. Start RetroArch.
2. Plug some USB stuff.
3. Don't profit.

**or**

1. Plug some USB stuff.
2. Start RetroArch.
3. Don't profit.

### Bisect Results

Discussed in the previous issue, came with the HID rewrite.

### Version/Commit

- RetroArch: 121ca3a plus #13308 from @gblues

### Environment information

- OS: Wii U, 5.5.5 PAL
- Compiler: official libretro-build-wiiu Docker image

**Comment by gblues** [comment]
I will note that the gamepad doesn't get formally disconnected--the "emergency shutdown" button combo (L+R+ZL+ZR+L3+R3) remains functional. For those affected by this issue, that's a bit faster recovery than power cycling the wiiu.

**Comment by gblues** [comment]
So, the way controller detection works is that we attempt to read the USB device ID, and then we compare that to our list of supported gamepads.

https://github.com/libretro/RetroArch/blame/master/input/connect/joypad_connection.c#L145-L173

Two things of note:
- the name is prioritized *over* VID/PID -- apparently there have been some issues where pads had the right name / wrong VID/PID and the name match was used to map them to compatible drivers
- the check is a *substring* check.

So, what I think is happening is that the Wii U is reading something super-generic like "USB Controller" as the device ID, and the substring search matches it against "Generic SNES USB Controller" and so we get the false match.

**Comment by vaguerant** [comment]
Nice work on #13321!

Sorry to bring the bad news, but now that I can actually use my keyboard without the GamePad dropping out, I am seeing some more keyboard-related issues.
- the keyboard works in the RetroArch menu, but not in-core, e.g. VICE x64 fast doesn't respond to keyboard input. I will bisect and see when this happened.
- performing a Main Menu -> Load Content then a Main Menu -> Quick Menu -> Close Content makes keyboard input stop working even in menus. Selecting Main Menu -> Restart RetroArch does bring keyboard input back.

**Comment by i30817** [comment]
Good old c substring bugs. Classic. The only bug i ever make a pr for retroarch was a 'hidden file' linux filter that was finding... any directory with a '.' because it was trying to search for the '.' substring.

**Comment by vaguerant** [comment]
Aha, thank you for that. Game Focus is indeed what's causing the only half my keys are working "issue" on older builds. I found that keyboards "just worked" in 1.9.0 stable, while from the 1.9.1 stable until b46daac3, Game Focus needs to be toggled. However, the current state of keyboards seems to be that they don't work in-core at all, Game Focus or not.

As long as we're messing with things, perhaps Game Focus should be configured to default to on for Wii U? Surely, nobody is plugging in their USB keyboard on Wii U to use *RetroArch hotkeys*. A far more sensible use case would be people wanting to use them to interact with the retro computer cores.

Will continue moving forward through commits to try to find where keyboards broke, although it's difficult due to the issues introduced *during* the HID rewrite which make it hard-to-impossible to test USB keyboards.

**Comment by i30817** [comment]
They honestly shouldn't be optional because then people would complain about 'buggy keyboards' if they forgot they had it on.

**Comment by vaguerant** [comment]
@gblues I just discovered something about keyboard support that should help in narrowing down what's going on. Keyboard support appears to work if you Load Content from the RetroArch menu, but not if you Start Core directly--supported in most (all?) home computer cores since they generally can do things other than read a disk/tape/etc. I'd say this makes the issue almost entirely resolved: keyboard support is already working except in the specific case that you're running a core without any content loaded.

This used to work (as of b46daac), but doesn't any longer. Is there anything in your HID rewrite which only becomes active when running via Load Content that Start Core is getting left out of?

**Comment by revvv** [comment]
The nightly builds now support USB gamepads again. To test it you just have to know a few things:

**tl;dr**
1. USB gamepads don't show up in "System Information", but appear in the input configuration menu.
2. You have to configure all buttons.
3. If the gamepad does not react, you have to replug it.

**Reasons**
1. autoconfig files are ignored, because you need to add `joypad_autoconfig_dir = "sd:/retroarch/autoconfig/hid"`
2. The Wii U autoconfig lookup only matches against the device name. VID/PID are not available in `task_autodetect.c`
    That's the reason why you always see `0/0` instead of the real VID/PID.
    The device names used for lookup are hardcoded in the various `input/connect/connect_*.c get_name()` functions.
    If you save an autoconfig file, it will contain this device name.
3. If you restart RetroArch with a valid autoconfig file, the gamepad will now show up in "System Information".

**Comment by LibretroAdmin** [comment]
Does this issue still happen or has it now been fixed?

**Comment by vaguerant** [comment]
Sorry for the late response.

The issue as originally filed no longer occurs, but the fix for the controller disconnection either exposed or caused a second issue, described in [these (1)](https://github.com/libretro/RetroArch/issues/13309#issuecomment-986013293) [comments (2)](https://github.com/libretro/RetroArch/issues/13309#issuecomment-997689663).

I'll recap the above comments here for better readability:

<hr />

USB HID keyboards are now broken when running cores contentless. Most of the PC cores for platforms like Commodore 64, 128, Amiga, etc. allow the user to select Start Core and run the system's BIOS or operating system rather than launching directly into content.

When selecting Start Core, keyboards stop working completely. They do not respond within the running core **or** in the RetroArch menu. Resolving this issue requires a full restart of RetroArch.

When running cores via Load Content, this issue does not occur.

Keyboards worked normally via Start Core as of b46daac3, which was before the HID rewrite, but as the HID rewrite broke keyboards for a while, it's not really possible to get a proper bisect of exactly when the issue started.

<hr />

I've also attached a log here of booting into VICE x64 fast with some content (a `.g64` disk image), using the keyboard normally, then using Close Content followed by Start Core to boot contentless, where the keyboard completely stops working. I don't see anything informative in the log, the HID initializing steps look the same to me and there's nothing alarming as far as I can tell.

[retroarch.log](https://github.com/libretro/RetroArch/files/9798356/retroarch.log)
